File: backend/services/GitHubService.js
/**
 * GitHub Service
 * Handles GitHub API operations for PR management
 */

import fetch from 'node-fetch';

export class GitHubService {
  constructor() {
    this.token = process.env.GITHUB_TOKEN;
    this.baseUrl = 'https://api.github.com';
    this.repos = process.env.GITHUB_REPOS ? process.env.GITHUB_REPOS.split(',') : [];
    
    if (!this.token) {
      console.warn('⚠️ GITHUB_TOKEN not set - GitHub integration disabled');
    }
  }

  /**
   * Check if GitHub is properly configured
   */
  isConfigured() {
    return !!this.token;
  }

  /**
   * Get headers for GitHub API requests
   */
  getHeaders() {
    return {
      'Authorization': `Bearer ${this.token}`,
      'Accept': 'application/vnd.github.v3+json',
      'Content-Type': 'application/json',
      'User-Agent': 'CodegenApp/1.0.0'
    };
  }

  /**
   * Get repository information
   * @param {string} owner - Repository owner
   * @param {string} repo - Repository name
   * @returns {Promise<Object>} Repository information
   */
  async getRepository(owner, repo) {
    if (!this.isConfigured()) {
      throw new Error('GitHub not configured. Please set GITHUB_TOKEN');
    }

    try {
      console.log(`🐙 Getting repository info: ${owner}/${repo}`);

      const response = await fetch(`${this.baseUrl}/repos/${owner}/${repo}`, {
        method: 'GET',
        headers: this.getHeaders()
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorData}`);
      }

      const repoData = await response.json();
      console.log(`✅ Repository info retrieved: ${repoData.full_name}`);

      return {
        id: repoData.id,
        name: repoData.name,
        full_name: repoData.full_name,
        owner: repoData.owner.login,
        default_branch: repoData.default_branch,
        clone_url: repoData.clone_url,
        ssh_url: repoData.ssh_url,
        html_url: repoData.html_url,
        private: repoData.private,
        created_at: repoData.created_at,
        updated_at: repoData.updated_at
      };

    } catch (error) {
      console.error('❌ GitHub repository error:', error.message);
      throw new Error(`Failed to get repository: ${error.message}`);
    }
  }

  /**
   * Create a pull request
   * @param {Object} params - PR parameters
   * @param {string} params.owner - Repository owner
   * @param {string} params.repo - Repository name
   * @param {string} params.title - PR title
   * @param {string} params.body - PR body
   * @param {string} params.head - Head branch
   * @param {string} params.base - Base branch
   * @returns {Promise<Object>} Created PR information
   */
  async createPullRequest({ owner, repo, title, body, head, base = 'main' }) {
    if (!this.isConfigured()) {
      throw new Error('GitHub not configured. Please set GITHUB_TOKEN');
    }

    try {
      console.log(`🐙 Creating PR: ${owner}/${repo} - ${title}`);

      const payload = {
        title,
        body,
        head,
        base,
        maintainer_can_modify: true
      };

      const response = await fetch(`${this.baseUrl}/repos/${owner}/${repo}/pulls`, {
        method: 'POST',
        headers: this.getHeaders(),
        body: JSON.stringify(payload)
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorData}`);
      }

      const prData = await response.json();
      console.log(`✅ PR created successfully: #${prData.number}`);

      return {
        id: prData.id,
        number: prData.number,
        title: prData.title,
        body: prData.body,
        state: prData.state,
        html_url: prData.html_url,
        head: {
          ref: prData.head.ref,
          sha: prData.head.sha
        },
        base: {
          ref: prData.base.ref,
          sha: prData.base.sha
        },
        created_at: prData.created_at,
        updated_at: prData.updated_at,
        mergeable: prData.mergeable,
        merged: prData.merged
      };

    } catch (error) {
      console.error('❌ GitHub PR creation error:', error.message);
      throw new Error(`Failed to create PR: ${error.message}`);
    }
  }

  /**
   * Get pull request information
   * @param {string} owner - Repository owner
   * @param {string} repo - Repository name
   * @param {number} prNumber - PR number
   * @returns {Promise<Object>} PR information
   */
  async getPullRequest(owner, repo, prNumber) {
    if (!this.isConfigured()) {
      throw new Error('GitHub not configured. Please set GITHUB_TOKEN');
    }

    try {
      console.log(`🐙 Getting PR info: ${owner}/${repo}#${prNumber}`);

      const response = await fetch(`${this.baseUrl}/repos/${owner}/${repo}/pulls/${prNumber}`, {
        method: 'GET',
        headers: this.getHeaders()
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorData}`);
      }

      const prData = await response.json();
      console.log(`✅ PR info retrieved: #${prData.number} - ${prData.state}`);

      return {
        id: prData.id,
        number: prData.number,
        title: prData.title,
        body: prData.body,
        state: prData.state,
        html_url: prData.html_url,
        head: {
          ref: prData.head.ref,
          sha: prData.head.sha
        },
        base: {
          ref: prData.base.ref,
          sha: prData.base.sha
        },
        created_at: prData.created_at,
        updated_at: prData.updated_at,
        mergeable: prData.mergeable,
        merged: prData.merged,
        merge_commit_sha: prData.merge_commit_sha
      };

    } catch (error) {
      console.error('❌ GitHub PR retrieval error:', error.message);
      throw new Error(`Failed to get PR: ${error.message}`);
    }
  }

  /**
   * Get PR status checks
   * @param {string} owner - Repository owner
   * @param {string} repo - Repository name
   * @param {string} sha - Commit SHA
   * @returns {Promise<Object>} Status checks information
   */
  async getPRStatusChecks(owner, repo, sha) {
    if (!this.isConfigured()) {
      throw new Error('GitHub not configured. Please set GITHUB_TOKEN');
    }

    try {
      console.log(`🐙 Getting status checks for: ${owner}/${repo}@${sha}`);

      const response = await fetch(`${this.baseUrl}/repos/${owner}/${repo}/commits/${sha}/status`, {
        method: 'GET',
        headers: this.getHeaders()
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorData}`);
      }

      const statusData = await response.json();
      console.log(`✅ Status checks retrieved: ${statusData.state}`);

      return {
        state: statusData.state, // pending, success, failure, error
        total_count: statusData.total_count,
        statuses: statusData.statuses.map(status => ({
          id: status.id,
          state: status.state,
          description: status.description,
          target_url: status.target_url,
          context: status.context,
          created_at: status.created_at,
          updated_at: status.updated_at
        })),
        sha: statusData.sha,
        commit_url: statusData.url
      };

    } catch (error) {
      console.error('❌ GitHub status checks error:', error.message);
      throw new Error(`Failed to get status checks: ${error.message}`);
    }
  }

  /**
   * Merge a pull request
   * @param {Object} params - Merge parameters
   * @param {string} params.owner - Repository owner
   * @param {string} params.repo - Repository name
   * @param {number} params.prNumber - PR number
   * @param {string} params.commitTitle - Merge commit title
   * @param {string} params.commitMessage - Merge commit message
   * @param {string} params.mergeMethod - Merge method (merge, squash, rebase)
   * @returns {Promise<Object>} Merge result
   */
  async mergePullRequest({ owner, repo, prNumber, commitTitle, commitMessage, mergeMethod = 'merge' }) {
    if (!this.isConfigured()) {
      throw new Error('GitHub not configured. Please set GITHUB_TOKEN');
    }

    try {
      console.log(`🐙 Merging PR: ${owner}/${repo}#${prNumber}`);

      const payload = {
        commit_title: commitTitle,
        commit_message: commitMessage,
        merge_method: mergeMethod
      };

      const response = await fetch(`${this.baseUrl}/repos/${owner}/${repo}/pulls/${prNumber}/merge`, {
        method: 'PUT',
        headers: this.getHeaders(),
        body: JSON.stringify(payload)
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorData}`);
      }

      const mergeData = await response.json();
      console.log(`✅ PR merged successfully: ${mergeData.sha}`);

      return {
        sha: mergeData.sha,
        merged: mergeData.merged,
        message: mergeData.message,
        merge_method: mergeMethod,
        merged_at: new Date().toISOString()
      };

    } catch (error) {
      console.error('❌ GitHub PR merge error:', error.message);
      throw new Error(`Failed to merge PR: ${error.message}`);
    }
  }

  /**
   * Add comment to pull request
   * @param {Object} params - Comment parameters
   * @param {string} params.owner - Repository owner
   * @param {string} params.repo - Repository name
   * @param {number} params.prNumber - PR number
   * @param {string} params.body - Comment body
   * @returns {Promise<Object>} Comment information
   */
  async addPRComment({ owner, repo, prNumber, body }) {
    if (!this.isConfigured()) {
      throw new Error('GitHub not configured. Please set GITHUB_TOKEN');
    }

    try {
      console.log(`🐙 Adding comment to PR: ${owner}/${repo}#${prNumber}`);

      const payload = { body };

      const response = await fetch(`${this.baseUrl}/repos/${owner}/${repo}/issues/${prNumber}/comments`, {
        method: 'POST',
        headers: this.getHeaders(),
        body: JSON.stringify(payload)
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorData}`);
      }

      const commentData = await response.json();
      console.log(`✅ Comment added successfully: ${commentData.id}`);

      return {
        id: commentData.id,
        body: commentData.body,
        html_url: commentData.html_url,
        created_at: commentData.created_at,
        updated_at: commentData.updated_at
      };

    } catch (error) {
      console.error('❌ GitHub comment error:', error.message);
      throw new Error(`Failed to add comment: ${error.message}`);
    }
  }

  /**
   * Test GitHub API connection
   * @returns {Promise<Object>} Connection test result
   */
  async testConnection() {
    if (!this.isConfigured()) {
      return {
        success: false,
        error: 'GitHub not configured. Please set GITHUB_TOKEN'
      };
    }

    try {
      console.log(`🐙 Testing GitHub API connection`);

      const response = await fetch(`${this.baseUrl}/user`, {
        method: 'GET',
        headers: this.getHeaders()
      });

      if (!response.ok) {
        throw new Error(`GitHub API test failed: ${response.status}`);
      }

      const userData = await response.json();
      console.log(`✅ GitHub API connection successful: ${userData.login}`);

      return {
        success: true,
        user: {
          login: userData.login,
          id: userData.id,
          name: userData.name,
          email: userData.email
        },
        rate_limit: {
          limit: response.headers.get('x-ratelimit-limit'),
          remaining: response.headers.get('x-ratelimit-remaining'),
          reset: response.headers.get('x-ratelimit-reset')
        },
        timestamp: new Date().toISOString()
      };

    } catch (error) {
      console.error('❌ GitHub connection test failed:', error.message);
      return {
        success: false,
        error: error.message,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Get service status
   */
  getStatus() {
    return {
      service: 'GitHub',
      configured: this.isConfigured(),
      token_set: !!this.token,
      base_url: this.baseUrl,
      repos: this.repos
    };
  }
}

File: backend/app/services/github_service.py
"""
GitHub Service

This service provides GitHub API integration for repository management,
webhook handling, and PR operations.
"""

import logging
import aiohttp
import json
from typing import Dict, Any, List, Optional
from urllib.parse import urljoin

from app.config.settings import get_settings

logger = logging.getLogger(__name__)


class GitHubRepository:
    """GitHub repository model"""
    
    def __init__(self, data: Dict[str, Any]):
        self.id = data.get("id")
        self.name = data.get("name")
        self.full_name = data.get("full_name")
        self.description = data.get("description")
        self.private = data.get("private", False)
        self.clone_url = data.get("clone_url")
        self.ssh_url = data.get("ssh_url")
        self.default_branch = data.get("default_branch", "main")
        self.owner = data.get("owner", {})


class GitHubPullRequest:
    """GitHub pull request model"""
    
    def __init__(self, data: Dict[str, Any]):
        self.id = data.get("id")
        self.number = data.get("number")
        self.title = data.get("title")
        self.body = data.get("body")
        self.state = data.get("state")
        self.head = data.get("head", {})
        self.base = data.get("base", {})
        self.mergeable = data.get("mergeable")
        self.merged = data.get("merged", False)
        self.html_url = data.get("html_url")


class GitHubService:
    """Service for GitHub API operations"""
    
    def __init__(self):
        self.settings = get_settings()
        self.github_token = self.settings.github_token
        self.base_url = "https://api.github.com"
        
    async def get_repositories(self, org: Optional[str] = None) -> List[GitHubRepository]:
        """
        Get repositories for the authenticated user or organization
        
        Args:
            org: Organization name (optional)
            
        Returns:
            List of GitHubRepository objects
        """
        try:
            if org:
                url = f"{self.base_url}/orgs/{org}/repos"
            else:
                url = f"{self.base_url}/user/repos"
            
            headers = {
                "Authorization": f"token {self.github_token}",
                "Accept": "application/vnd.github.v3+json"
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url, headers=headers) as response:
                    if response.status == 200:
                        repos_data = await response.json()
                        return [GitHubRepository(repo) for repo in repos_data]
                    else:
                        error_text = await response.text()
                        logger.error(f"Failed to get repositories: {response.status} - {error_text}")
                        return []
                        
        except Exception as e:
            logger.error(f"Error getting repositories: {e}")
            return []
    
    async def get_repository(self, owner: str, repo: str) -> Optional[GitHubRepository]:
        """
        Get a specific repository
        
        Args:
            owner: Repository owner
            repo: Repository name
            
        Returns:
            GitHubRepository object or None
        """
        try:
            url = f"{self.base_url}/repos/{owner}/{repo}"
            headers = {
                "Authorization": f"token {self.github_token}",
                "Accept": "application/vnd.github.v3+json"
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url, headers=headers) as response:
                    if response.status == 200:
                        repo_data = await response.json()
                        return GitHubRepository(repo_data)
                    else:
                        logger.error(f"Repository not found: {owner}/{repo}")
                        return None
                        
        except Exception as e:
            logger.error(f"Error getting repository {owner}/{repo}: {e}")
            return None
    
    async def set_webhook(
        self, 
        owner: str, 
        repo: str, 
        webhook_url: str,
        events: List[str] = None
    ) -> bool:
        """
        Set up a webhook for a repository
        
        Args:
            owner: Repository owner
            repo: Repository name
            webhook_url: URL to receive webhook notifications
            events: List of events to subscribe to
            
        Returns:
            True if successful, False otherwise
        """
        if events is None:
            events = ["pull_request", "push", "issues"]
        
        try:
            url = f"{self.base_url}/repos/{owner}/{repo}/hooks"
            headers = {
                "Authorization": f"token {self.github_token}",
                "Accept": "application/vnd.github.v3+json",
                "Content-Type": "application/json"
            }
            
            webhook_config = {
                "name": "web",
                "active": True,
                "events": events,
                "config": {
                    "url": webhook_url,
                    "content_type": "json",
                    "insecure_ssl": "0"
                }
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.post(url, headers=headers, json=webhook_config) as response:
                    if response.status == 201:
                        logger.info(f"Webhook set successfully for {owner}/{repo}")
                        return True
                    else:
                        error_text = await response.text()
                        logger.error(f"Failed to set webhook: {response.status} - {error_text}")
                        return False
                        
        except Exception as e:
            logger.error(f"Error setting webhook for {owner}/{repo}: {e}")
            return False
    
    async def get_pull_request(
        self, 
        owner: str, 
        repo: str, 
        pr_number: int
    ) -> Optional[GitHubPullRequest]:
        """
        Get a specific pull request
        
        Args:
            owner: Repository owner
            repo: Repository name
            pr_number: Pull request number
            
        Returns:
            GitHubPullRequest object or None
        """
        try:
            url = f"{self.base_url}/repos/{owner}/{repo}/pulls/{pr_number}"
            headers = {
                "Authorization": f"token {self.github_token}",
                "Accept": "application/vnd.github.v3+json"
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url, headers=headers) as response:
                    if response.status == 200:
                        pr_data = await response.json()
                        return GitHubPullRequest(pr_data)
                    else:
                        logger.error(f"Pull request not found: {owner}/{repo}#{pr_number}")
                        return None
                        
        except Exception as e:
            logger.error(f"Error getting pull request {owner}/{repo}#{pr_number}: {e}")
            return None
    
    async def merge_pull_request(
        self,
        owner: str,
        repo: str,
        pr_number: int,
        commit_title: Optional[str] = None,
        commit_message: Optional[str] = None,
        merge_method: str = "merge"
    ) -> bool:
        """
        Merge a pull request
        
        Args:
            owner: Repository owner
            repo: Repository name
            pr_number: Pull request number
            commit_title: Title for the merge commit
            commit_message: Message for the merge commit
            merge_method: Merge method (merge, squash, rebase)
            
        Returns:
            True if successful, False otherwise
        """
        try:
            url = f"{self.base_url}/repos/{owner}/{repo}/pulls/{pr_number}/merge"
            headers = {
                "Authorization": f"token {self.github_token}",
                "Accept": "application/vnd.github.v3+json",
                "Content-Type": "application/json"
            }
            
            merge_data = {
                "merge_method": merge_method
            }
            
            if commit_title:
                merge_data["commit_title"] = commit_title
            if commit_message:
                merge_data["commit_message"] = commit_message
            
            async with aiohttp.ClientSession() as session:
                async with session.put(url, headers=headers, json=merge_data) as response:
                    if response.status == 200:
                        logger.info(f"Pull request merged successfully: {owner}/{repo}#{pr_number}")
                        return True
                    else:
                        error_text = await response.text()
                        logger.error(f"Failed to merge pull request: {response.status} - {error_text}")
                        return False
                        
        except Exception as e:
            logger.error(f"Error merging pull request {owner}/{repo}#{pr_number}: {e}")
            return False
    
    async def create_issue_comment(
        self,
        owner: str,
        repo: str,
        issue_number: int,
        comment: str
    ) -> bool:
        """
        Create a comment on an issue or pull request
        
        Args:
            owner: Repository owner
            repo: Repository name
            issue_number: Issue or PR number
            comment: Comment text
            
        Returns:
            True if successful, False otherwise
        """
        try:
            url = f"{self.base_url}/repos/{owner}/{repo}/issues/{issue_number}/comments"
            headers = {
                "Authorization": f"token {self.github_token}",
                "Accept": "application/vnd.github.v3+json",
                "Content-Type": "application/json"
            }
            
            comment_data = {"body": comment}
            
            async with aiohttp.ClientSession() as session:
                async with session.post(url, headers=headers, json=comment_data) as response:
                    if response.status == 201:
                        logger.info(f"Comment created successfully on {owner}/{repo}#{issue_number}")
                        return True
                    else:
                        error_text = await response.text()
                        logger.error(f"Failed to create comment: {response.status} - {error_text}")
                        return False
                        
        except Exception as e:
            logger.error(f"Error creating comment on {owner}/{repo}#{issue_number}: {e}")
            return False
    
    async def get_branches(self, owner: str, repo: str) -> List[str]:
        """
        Get all branches for a repository
        
        Args:
            owner: Repository owner
            repo: Repository name
            
        Returns:
            List of branch names
        """
        try:
            url = f"{self.base_url}/repos/{owner}/{repo}/branches"
            headers = {
                "Authorization": f"token {self.github_token}",
                "Accept": "application/vnd.github.v3+json"
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url, headers=headers) as response:
                    if response.status == 200:
                        branches_data = await response.json()
                        return [branch["name"] for branch in branches_data]
                    else:
                        logger.error(f"Failed to get branches for {owner}/{repo}")
                        return []
                        
        except Exception as e:
            logger.error(f"Error getting branches for {owner}/{repo}: {e}")
            return []
